From 5e22888b7cd7ef348f905627de7fece18bb94de7 Mon Sep 17 00:00:00 2001
From: Mark Horvath <mark.horvath@arm.com>
Date: Fri, 12 Nov 2021 12:22:00 +0100
Subject: [PATCH 1/1] Chunk writing added for otaPal_WriteBlock

If file block size is bigger then PSA_FWU_MAX_BLOCK_SIZE flash is
written in chunks.

Signed-off-by: Mark Horvath <mark.horvath@arm.com>
---
 ota_pal.c | 21 +++++++++++++++++++--
 ota_pal.h |  2 +-
 2 files changed, 20 insertions(+), 3 deletions(-)

diff --git a/ota_pal.c b/ota_pal.c
index 40654a8..7659b62 100644
--- a/ota_pal.c
+++ b/ota_pal.c
@@ -295,19 +295,36 @@ OtaPalStatus_t otaPal_CloseFile( OtaFileContext_t * const pFileContext )
  */
 int16_t otaPal_WriteBlock( OtaFileContext_t * const pFileContext,
                            uint32_t ulOffset,
-                           uint8_t * const pcData,
+                           const uint8_t * pcData,
                            uint32_t ulBlockSize )
 {
+    uint32_t remaining_blocksize = ulBlockSize;
+
     if( (pFileContext == NULL) || (pFileContext != pxSystemContext ) || ( xOTAImageID == TFM_FWU_INVALID_IMAGE_ID ) )
     {
         return -1;
     }
 
     /* Call the TF-M Firmware Update service to write image data. */
+    while(remaining_blocksize > PSA_FWU_MAX_BLOCK_SIZE)
+    {
+        if( psa_fwu_write( ( psa_image_id_t ) xOTAImageID,
+                           ( size_t ) ulOffset,
+                           ( const void * )pcData,
+                           PSA_FWU_MAX_BLOCK_SIZE ) != PSA_SUCCESS )
+        {
+            return -1;
+        }
+
+        ulOffset += PSA_FWU_MAX_BLOCK_SIZE;
+        pcData += PSA_FWU_MAX_BLOCK_SIZE;
+        remaining_blocksize -= PSA_FWU_MAX_BLOCK_SIZE;
+    }
+
     if( psa_fwu_write( ( psa_image_id_t ) xOTAImageID,
                        ( size_t ) ulOffset,
                        ( const void * )pcData,
-                       ( size_t ) ulBlockSize ) != PSA_SUCCESS )
+                       ( size_t ) remaining_blocksize ) != PSA_SUCCESS )
     {
         return -1;
     }
diff --git a/ota_pal.h b/ota_pal.h
index c837b96..7a0fcf2 100644
--- a/ota_pal.h
+++ b/ota_pal.h
@@ -136,7 +136,7 @@ OtaPalStatus_t otaPal_CloseFile( OtaFileContext_t * const pFileContext );
  */
 int16_t otaPal_WriteBlock( OtaFileContext_t * const pFileContext,
                            uint32_t ulOffset,
-                           uint8_t * const pData,
+                           const uint8_t * pData,
                            uint32_t ulBlockSize );
 
 /**
-- 
2.25.1

